app/src/main/java/com/practicum/playlistmaker/SearchActivity.kt

```diff
@@ -47,41 +75,103 @@ class SearchActivity : AppCompatActivity() {
         }
         searchTextEdit.addTextChangedListener(searchTextWatcher)
 
-        val searchTextEditLayout = findViewById<TextInputLayout>(R.id.search_edit_text_layout)
+        searchTextEdit.setOnEditorActionListener { _, actionId, _ ->
+            if (actionId == EditorInfo.IME_ACTION_DONE) {
+                songSearch()
+                searchTextEdit.clearFocus()
+                true
+            }
+            false
+        }
+
+        //Hide keyboard when clear button clicked
         searchTextEditLayout.setEndIconOnClickListener {
             searchTextEdit.setText(SEARCH_STRING_DEFAULT)
+            trackList.clear()
+            adapter.notifyDataSetChanged()
+
             val inputMethodManager =
                 getSystemService(Context.INPUT_METHOD_SERVICE) as? InputMethodManager
             inputMethodManager?.hideSoftInputFromWindow(searchTextEdit.windowToken, 0)
             searchTextEdit.clearFocus()
         }
 
         val recycler = findViewById<RecyclerView>(R.id.track_recycler_view)
-        recycler.adapter = TrackAdapter(TrackListMock.create())
+        recycler.adapter = adapter
 
     }
 
     fun setCleanSearchButtonVisibility() {
-        val searchTextEditLayout = findViewById<TextInputLayout>(R.id.search_edit_text_layout)
         searchTextEditLayout.isEndIconVisible = !searchString.isEmpty()
     }
 
     override fun onSaveInstanceState(outState: Bundle) {
         super.onSaveInstanceState(outState)
         outState.putString(SEARCH_STRING, searchString)
+        outState.putInt(CONNECTION_ERROR_VISIBLE, connectionErrorFrame.visibility)
+        outState.putInt(NOTHING_FOUND_VISIBLE, nothingFoundFrame.visibility)
+        outState.putString(TRACK_LIST, gson.toJson(Tracks(trackList)))
     }
 
     override fun onRestoreInstanceState(savedInstanceState: Bundle) {
         super.onRestoreInstanceState(savedInstanceState)
         searchString = savedInstanceState.getString(SEARCH_STRING, SEARCH_STRING_DEFAULT)
             ?: SEARCH_STRING_DEFAULT
-        val searchTextEdit = findViewById<TextInputEditText>(R.id.search_edit_text)
         searchTextEdit.setText(searchString)
+
+        trackList.clear()
+        trackList.addAll(gson.fromJson(savedInstanceState.getString(TRACK_LIST), Tracks::class.java).list)
+        adapter.notifyDataSetChanged()
+        connectionErrorFrame.visibility = savedInstanceState.getInt(CONNECTION_ERROR_VISIBLE, View.GONE)
+        nothingFoundFrame.visibility = savedInstanceState.getInt(NOTHING_FOUND_VISIBLE, View.GONE)
+
+    }
+
+    private fun songSearch() {
+        itunesService.songSearch(searchString)
+            .enqueue(object : Callback<SongSearchResponse> {
+                override fun onResponse(
+                    call: Call<SongSearchResponse>,
+                    response: Response<SongSearchResponse>
+                ) {
+                    val searchResponse = response.body()
+
+                    if (response.code() == HTML_OK) {
```

Comment:
```suggestion
                    if (response.isSuccessfull) {
```
Таким образом, тебе не придётся хардкодить число 200 в своём коде и проверка будет более валидной, потому что она проверяет, что код находится в промежутке 200..300